server-utils/src/reactor.rs

```diff
@@ -1,68 +1,82 @@
-//! Event Loop Remote
+//! Event Loop Executor
 //! Either spawns a new event loop, or re-uses provided one.
 
 use std::{io, thread};
 use std::sync::mpsc;
-use tokio_core;
+use tokio;
 
 use core::futures::{self, Future};
 
-/// Possibly uninitialized event loop remote.
+/// Thread pool thread count.
+//
+// NOTE: It may be better to use `num_cpus` and dynamically choose pool size
+// instead.
+const THREAD_POOL_SIZE: usize = 1;
+
+/// Possibly uninitialized event loop executor.
 #[derive(Debug)]
-pub enum UninitializedRemote {
-	/// Shared instance of remote.
-	Shared(tokio_core::reactor::Remote),
+pub enum UninitializedExecutor {
+	/// Shared instance of executor.
+	Shared(tokio::runtime::TaskExecutor),
 	/// Event Loop should be spawned by the transport.
 	Unspawned,
 }
 
-impl UninitializedRemote {
-	/// Initializes remote.
-	/// In case there is no shared remote, will spawn a new event loop.
-	/// Dropping `Remote` closes the loop.
-	pub fn initialize(self) -> io::Result<Remote> {
+impl UninitializedExecutor {
+	/// Initializes executor.
+	/// In case there is no shared executor, will spawn a new event loop.
+	/// Dropping `Executor` closes the loop.
+	pub fn initialize(self) -> io::Result<Executor> {
 		self.init_with_name("event.loop")
 	}
 
-	/// Initializes remote.
-	/// In case there is no shared remote, will spawn a new event loop.
-	/// Dropping `Remote` closes the loop.
-	pub fn init_with_name<T: Into<String>>(self, name: T) -> io::Result<Remote> {
+	/// Initializes executor.
+	/// In case there is no shared executor, will spawn a new event loop.
+	/// Dropping `Executor` closes the loop.
+	pub fn init_with_name<T: Into<String>>(self, name: T) -> io::Result<Executor> {
 		match self {
-			UninitializedRemote::Shared(remote) => Ok(Remote::Shared(remote)),
-			UninitializedRemote::Unspawned => RpcEventLoop::with_name(Some(name.into())).map(Remote::Spawned),
+			UninitializedExecutor::Shared(executor) => Ok(Executor::Shared(executor)),
+			UninitializedExecutor::Unspawned => RpcEventLoop::with_name(Some(name.into())).map(Executor::Spawned),
 		}
 	}
 }
 
-/// Initialized Remote
+/// Initialized Executor
 #[derive(Debug)]
-pub enum Remote {
+pub enum Executor {
 	/// Shared instance
-	Shared(tokio_core::reactor::Remote),
+	Shared(tokio::runtime::TaskExecutor),
 	/// Spawned Event Loop
 	Spawned(RpcEventLoop),
 }
 
-impl Remote {
-	/// Get remote associated with this event loop.
-	pub fn remote(&self) -> tokio_core::reactor::Remote {
+impl Executor {
+	/// Get tokio executor associated with this event loop.
+	pub fn executor(&self) -> tokio::runtime::TaskExecutor {
 		match *self {
-			Remote::Shared(ref remote) => remote.clone(),
-			Remote::Spawned(ref eloop) => eloop.remote(),
+			Executor::Shared(ref executor) => executor.clone(),
+			Executor::Spawned(ref eloop) => eloop.executor(),
```

Comment:
Mayhaps instead of having `Spawned` variant we could use default tokio runtime?
The whole point of having `Executor/EventLoop` abstraction was to provide the users a simple way to start and stop the server without manually spawning some additional threads - with new `tokio` I suppose that's the job that `tokio::runtime` is fulfiling.

Reply:
True but one reason to keep it as is would be if some kind of customization may be needed. For example there may be certain things which you want to spawn on the thread pool, and others you want to spawn on a specific thread, to separate tasks that need to do work from those that only need to be run to completion. You can also do more advanced things like little optimizations for things you might `Box` in bulk. I use a custom runtime in a few of my projects for those and other reasons, having the flexibility to add it later without many changes is nice.

Regardless of that, it now looks now like I'll have to add a `tokio::reactor::Handle` back to the executor-remote anyway due to some recent changes I had to make to `parity-tokio-ipc`, et al. so the variant will probably need to stick around for now.

Reply:
Also note that in a pure-futures library, everything would be wrapped up in a future and the runtime would just be run on the main thread. There would be no built-in event loop, it would be left to the user. Once futures starts to stabilize and becomes more ubiquitous in the Parity ecosystem it will be worth redesigning jsonrpc in that direction (and I'll be doing my best to move things in that direction sooner rather than later).